man/setres.8
.TH "setres" 8
.SH "NAME"
setres \- Create or modify a cobalt scheduler reservation
.SH "SYNOPSIS"
.BR setres
\fB-n\ \fIname\fR\ [\fB-m\fR]\ [\fB-A\ \fIproject\fR]\ [\fB-c\fR]
[\fB--allow_passthrough\fR] [\fB--block_passthrough\fR]\ [\fB-D\fR]
[\fB-d \ \fIduration\fR]\ [\fB--debug\fR] [\fB-p \ \fIlocation\fR]
[\fB-q\ \fIqueue\fR]\ [\fB-s\ \fIstarttime\fR] [\fB-u\ \fIuser-list\fR]
[\fIreservation locations\fR]
.TP
.B setres
[\fB--res_id\ \fIid\fR]\ [\fB--cycle_id\ \fIid\fR]\ [\fB--force\fR]
.TP
.BR setres\ -h
.TP
.BR setres\ --version

.SH "DESCRIPTION"
Creates or modifies a scheduler reservation.  Reservation and cycle ids may also
be reset.

.SH "OPTIONS"
.TP
.B \-A \-\-project \fIproject
Set project name to associate with the reservation.
.TP
.B \-\-allow_passthrough
Allow pass through connection on systems with interconnects that allow
passthrough communication.
.TP
.B \-\-block_passthrough
Block pass through connections on systems with interconnects that allow
passthrough communication.
.TP
.B \-c \-\-cycletime \fItime
Set the cycle time (in minutes or DD:HH:MM:SS).  This is the amount of
time from reservation start until it is automatically renewed.  This
can be used to create repeating reservations.
.TP
.B \-d \-\-duration \fIduration
Set duration (in minutes or HH:MM:SS)
.TP
.B \-D \-\-defer
Defer current (or next) iteration of recurring reservation. This must be used
with the
.B -m
flag.
.TP
.B \-\-debug
Turn on communication debugging.
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-m \-\-modify
Modify an existing reservation. The target reservation specified with
.BR -n .
.TP
.B \-n \-\-name
Set reservation name.  Names must be unique for all pending and active
reservations on a system.
.TP
.B \-p \-\-partition \fIlocation
Set the location to use for a reservation.  This may be used instead of positional arguments
for locations.  All locations in a reservation must exist and must
be managed by the system component at the time the reservation is set and active.
.TP
.B \-q \-\-queue \fIqueue
Set the queue name. Optional. Queues may already exist and have jobs in them.
Jobs currently running in a target queue will not be affected by applying a
reservation to the queue.  Jobs that are queued in the target queue will not
start until the reservation becomes active.  Jobs in a reservations against an
existing queue will be permitted to run on all nodes in that queue.

If this option isn't specified, a queue "R.name" will be created where name is the
reservation name specified by the
.B -n
argument.
.TP
.B \-s \-\-starttime \fIstarttime
Set start time (supported formats include YYYY-MM-DD-HH:MM or
YYYY_MM_DD-HH:MM). The \fIstarttime\fR may also be "now," which will set the
reservation starttime to the current time and the reservation will immediately
activate for the next scheduling iteration.
.TP
.B \-u \-\-user \fIuser-list
Set user(s) for reservation. Set to "*" for all users. User names may be
provided as a colon-delimited (:) list.  User names must be valid on the node where
.BR setres (8)
is running.
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-\-cycle_id \fIid
Set the integer cycle id.  Without \-\-force_id this must be a larger value
than the current maximum cycle id.  This may not be used with any option other than
\-\-force_id
.TP
.B \-\-force_id
Only used with \-\-res_id or \-\-cycle_id options.  Will force the id generator
to start with the specified value.  Improper use of this option may cause
non-unique reservation ids and cycle ids to occur.
.TP
.B \-\-res_id \fIid
Set the integer reservation id.  Without \-\-force this must be a larger value
than the current maximum reservation id. This may not be used with any option other than
\-\-force_id

.SH "NOTES"
At a minimum all reservation creation requires use of the
.B -n, -s
and
.B -d
flags.  Partions and nodes must be specified as positional arguments or via the
.B -p
flag.
.P
On Cray systems nodes are referenced by their integer node id.  Nodes may be
specified as a comma-delimited list.  Ranges of node id's may be compacted with a
hyphen to an inclusive range, i.e. 1-4 will expand to 1,2,3,4.

.SH "SEE ALSO"
.BR showres (1),
.BR releaseres (8)
.SH "BUGS"


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 1.0.8 ==
 * Fixed a bug where a node going down during a running job could allow that
   node to be selected for draining.
 * Expanded a buffer used by the forkers that should significantly improve the
   speed of reading large ALPS responses and no longer interfere with node
   reacquisition
 * Cleaned up stray print statements in node information
 * Fixed an error where score could be set to a string rather than a float via
   the filter scripts
 * Fixed an error where the queue-manager statefile could be lost if Cobalt was
   brought down with a job in the "Terminal" state.

== Changes to 1.0.6 ==
 * Backfill now supported on Cray systems.
 * Major performance improvement to nodelist and nodeadm -l.
 * Fix for an error that can cause loss of job tracking during job startup
   resulting in a reported exit status of "1234567".
 * Fix for an error in system_script_forker that can cause an error loop
   if using stdout to string functionality.
 * Fix for an error that caused jobs with a "terminating" status to cause
   qstat to fail.

== Changes to 1.0.5 ==
 * Fix for an issue that would cause jobs with their location set to not
   run if they were submitted to a reservation.
 * Performance of setres improved significantly.
 * Fixed an issue where reservations and soft-down status on nodes may not
   be respected by job placement.
 * An enhanced_apkill script is now available to improve job cleanup and
   ensure that jobs that ignore signals are otherwise properly terminated.

== Changes to 1.0.4 ==
 * Fix for apid not being properly checked when apkill was being invoked.

== Changes to 1.0.3 ==
 * Fix for an issue for nodes coming out of disabled state and suddenly
   appearing in ALPS.
 * Interactive job support
 * Support for CAPMC memory management scripts added.  Cobalt now gracefully
   handles long startup times if these are being used.
 * Improved job cleanup using apkill.

== Changes to 1.0.2 ==
 * Singleton job filter (maxtotaljobs) now supported
 * Multiple alps_script_forkers running simultaneously is now supported. Forkers
   are chosen for execution based on which forker has the lowest number of jobs
   currently assigned to it.

== Changes to 1.0.0 ==
 * Initial support of Cray systems

== Changes to 0.99.42 ==
 * Update to database writer to handle attrs fields that are
   improperly formatted.

== Changes to 0.99.41 ==
 * Minor patch to clean up an issue where the user was not being properly
   cleared from BlueGene blocks at the end of a run.

== Changes to 0.99.40 ==
 * A nokillpg attribute may be added to a job to cause kill instead of killpg
   to be used for job termination so that a signal handler in a script may
   handle signal propigation itself.
 * Fixed a bug where a job requesting a location could cause Cobalt to drain
   for a job that would require the use of offline hardware
 * Added a check to the job validator where a job is rejected if the job size
   and block geometry have a mismatch.

== Changes to 0.99.39 ==
 * Session information added to job submission information.  This also includes
   submitting host information (Cobalt Ticket #4315)
 * A fix for a bug that can cause "phantom drains" to occur on lightly loaded
   cluster systems. (Cobalt Ticket #4319)

== Changes to 0.99.38 ==
 * Added numerous dependency changes.  Notably, jobs in dep_hold states will no
   longer accrue eligible wait time or score.  Previously, dep_hold jobs would
   accrue eligible wait times unlike admin_hold or user_hold states.
   (Cobalt Ticket #4308)
 * Two display fixes for cluster systems that allow multiple node-name prefixes
   to be handled correctly by the nodelist merging code  and to correctly
   compact names if leading 0's are used in location numbers like
   "foo001.machine". (Cobalt Ticket #4313 and #4277)
 * Corrected two race conditions in the BlueGene/Q system components where a
   location with offline hardware could be selected for draining on a
   scheduling pass.  Though Cobalt would never try to actually run on the
   offline block, It could end up draining resources needlessly, and impact
   machine usage.  The first could happen at any time, the second only occurred
   at the end of booting a block. (Cobalt Ticket #4311)

== Changes to 0.99.37 ==
 * Fixed issue where a reservation could hit it's activation check while it is being
   deactivated, causing those events should be reversed.  Released reservations can't
   activate. (Cobalt Ticket #4289)
 * COBALT_JOBSIZE and COBALT_PARTSIZE have been added to interactive jobs.
   (Cobalt Ticket #4294)
 * Job dependencies require the same project to gain score from eachother.
   (Cobalt Ticket #4293)
 * prologue_helper.py now handles "=" in arguments correctly. (Cobalt Ticket #4292)

== Changes to 0.99.36 ==
 * Fix for issues in get-bootable-blocks and with block booting.  Block names are now
   properly sorted.  A race condition between boot-block and update_block_state has
   been fixed.  The return statuses for boot-block have also been corrected for failed
   boots.  (Cobalt Ticket #4282 and #4280)

== Changes to 0.99.35 ==
 * Fixed an issue where a block being removed in the BGQ control system between block
   detection and the block being added for tracking could result in blocks being
   erroniously marked as idle. (Cobalt Ticket #4268)

== Changes to 0.99.34 ==
 * Fixed an issue where Cobalt could have a memory leak if boot-block.py was
   terminated early the boot tracking object may never be cleaned up.
 * Numerous enhancements and fixes to handle multiple overlaping queues in the
   system with respect to draining and backfilling. (Cobalt Ticket #663)
 * Fixed some typos in the documentation for qsub.
 * COBALT_RESID is now appropriately set in interactive mode jobs (Cobalt Ticket #703)

== Changes to 0.99.33 ==
 * Fixed issue where adding blocks while the system component is running can

src/clients/setres.py
#!/usr/bin/env python
"""
Setup reservations in the scheduler

Usage: %prog --help
Usage: %prog [options] <partition1> ... <partitionN>
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

'-A','--project',dest='project',type='string',help='set project name for partitions in args'
'-D','--defer',dest='defer',help='defer flag. needs to be used with -m',action='store_true'
'-c','--cycletime',dest='cycle',type='string',help='set cycle time in minutes or HH:MM:SS',callback=cb_time
'-d','--duration',dest='duration',type='string',help='duration time in minutes or HH:MM:SS',callback=cb_time
'-m','--modify_res',dest='modify_res',help='modify current reservation specified in -n',action='store_true'
'-n','--name',dest='name',type='string',help='reservation name to add or modify'
'-p','--partitions',dest='partitions',type='string',help='partition(s) (now optional) part1:..:partN'
'-q','--queue',dest='queue',type='string',help='queue name'
'-s','--starttime',dest='start',type='string',help='start date time: YYYY_MM_DD-HH:MM, or now',callback=cb_date
'-u','--user',dest='users',type='string',help='user id list (user1:user2:... or "*" for all users)',callback=cb_res_users
'--debug',dest='debug',help='turn on communication debugging',callback=cb_debug


'--allow_passthrough',dest='block_passthrough',help='allow passthrough',callback=cb_passthrough
'--block_passthrough',dest='block_passthrough',help='do not allow passthrough',callback=cb_passthrough

OPTIONS TO CHANGE IDS ONLY:

'--cycle_id',dest='cycle_id',type='int',help='cycle id'
'--force_id',dest='force_id',default=False,help='force id flag (only used with --cycle_id or --res_id',action='store_true'
'--res_id',dest='res_id',type='int',help='res id'
"""
import logging
import math
import sys
import time
from Cobalt import client_utils
from Cobalt.client_utils import cb_debug, cb_time, cb_date, cb_passthrough, cb_res_users
from Cobalt.Util import expand_num_list, compact_num_list
from Cobalt.arg_parser import ArgParse

__revision__ = '$Id: setres.py 2154 2011-05-25 00:22:56Z richp $'
__version__ = '$Version$'

SCHMGR = client_utils.SCHMGR
SYSMGR = client_utils.SYSMGR

def validate_starttime(parser):
    """
    make sure the start time plus duration is valid
    """
    _localtime = client_utils.parse_datetime(client_utils.cobalt_date(time.localtime(time.time())))
    _res_st    = parser.options.start + parser.options.duration
    if _res_st < _localtime:
        client_utils.logger.error("Start time + duration %s already in the pass", client_utils.sec_to_str(_res_st))
        sys.exit(1)

def set_res_id(parser):
    """
    set res id
    """
    if parser.options.force_id:
        client_utils.component_call(SCHMGR, False, 'force_res_id', (parser.options.res_id,))
        client_utils.logger.info("WARNING: Forcing res id to %s" % parser.options.res_id)
    else:
        client_utils.component_call(SCHMGR, False, 'set_res_id', (parser.options.res_id,))
        client_utils.logger.info("Setting res id to %s" % parser.options.res_id)

def set_cycle_id(parser):
    """
    set res id
    """
    cycle_id = parser.options.cycle_id
    if parser.options.force_id:
        client_utils.component_call(SCHMGR, False, 'force_cycle_id', (cycle_id,))
        client_utils.logger.info("WARNING: Forcing cycle id to %s" % str(cycle_id))
    else:
        client_utils.component_call(SCHMGR, False, 'set_cycle_id', (cycle_id,))
        client_utils.logger.info("Setting cycle_id to %s" % str(cycle_id))

def verify_locations(partitions):
    """
    verify that partitions are valid
    """
    check_partitions = partitions
    system_type = client_utils.component_call(SYSMGR, False, 'get_implementation', ())
    # if we have args then verify the args (partitions)
    if system_type in ['alps_system']:
        # nodes come in as a compact list.  expand this.
        check_partitions = []
        # if we're not a compact list, convert to a compact list.  Get this,
        # ideally, in one call
        for num_list in partitions:
            check_partitions.extend(expand_num_list(num_list))
            test_parts = client_utils.component_call(SYSMGR, False,
            'verify_locations', (check_partitions,))
            # On Cray we will be a little looser to make setting reservations
            # easier.
            client_utils.logger.info('Got nodes: %s', compact_num_list(test_parts))
            sys.exit(1)
    for p in check_partitions:
        print "checking parts"
        test_parts = client_utils.component_call(SYSMGR, False,
                'verify_locations', (check_partitions,))
        if len(test_parts) != len(check_partitions):
            missing = [p for p in check_partitions if p not in test_parts]
            client_utils.logger.error("Missing partitions: %s" % (" ".join(missing)))
            sys.exit(1)

def validate_args(parser,spec,opt_count):
    """
    Validate setres arguments. Will return true if we want to continue processing options.
    """
    if parser.options.partitions != None:
        parser.args += [part for part in parser.options.partitions.split(':')]

    if parser.options.cycle_id != None or parser.options.res_id != None:
        only_id_change = True
        if not parser.no_args() or (opt_count != 0):
            client_utils.logger.error('No partition arguments or other options allowed with id change options')
            sys.exit(1)
    else:
        only_id_change = False

    if parser.options.force_id and not only_id_change:
        client_utils.logger.error("--force_id can only be used with --cycle_id and/or --res_id.")
        sys.exit(1)

    if only_id_change:

        # make the ID change and we are done with setres
        
        if parser.options.res_id != None:
            set_res_id(parser)
        if parser.options.cycle_id != None:
            set_cycle_id(parser)

        continue_processing_options = False # quit, setres is done

    else:

        if parser.options.name is None:
            client_utils.print_usage(parser)
            sys.exit(1)

        if parser.no_args() and (parser.options.modify_res == None):
            client_utils.logger.error("Must supply either -p with value or partitions as arguments")
            sys.exit(1)

        if parser.options.start == None and parser.options.modify_res == None:
            client_utils.logger.error("Must supply a start time for the reservation with -s")
            sys.exit(1)

        if parser.options.duration == None and parser.options.modify_res == None:
            client_utils.logger.error("Must supply a duration time for the reservation with -d")
            sys.exit(1)

        if parser.options.defer != None and (parser.options.start != None or parser.options.cycle != None):
            client_utils.logger.error("Cannot use -D while changing start or cycle time")
            sys.exit(1)

        if not parser.no_args():
            verify_locations(parser.args)

        # if we have command line arguments put them in spec
        if not parser.no_args(): spec['partitions'] = ":".join(parser.args)

        continue_processing_options = True # continue, setres is not done.

    return continue_processing_options

        
def modify_reservation(parser):
    """
    this will handle reservation modifications
    """
    query = [{'name':parser.options.name, 'start':'*', 'cycle':'*', 'duration':'*'}]
    res_list = client_utils.component_call(SCHMGR, False, 'get_reservations', (query,))
    if not res_list:
        client_utils.logger.error("cannot find reservation named '%s'" % parser.options.name)
        sys.exit(1)

    updates = {} # updates to reservation
    if parser.options.defer != None:
        res = res_list[0]

        if not res['cycle']:
            client_utils.logger.error("Cannot use -D on a non-cyclic reservation")
            sys.exit(1)

        start    = res['start']
        duration = res['duration']
        cycle    = float(res['cycle'])
        now      = time.time()
        periods  = math.floor((now - start)/cycle)

        if(periods < 0):
            start += cycle
        elif(now - start) % cycle < duration:
            start += (periods + 1) * cycle
        else:
            start += (periods + 2) * cycle

        newstart = time.strftime("%c", time.localtime(start))
        client_utils.logger.info("Setting new start time for for reservation '%s': %s" % (res['name'], newstart))

        updates['start'] = start

        #add a field to updates to indicate we're deferring:
        updates['defer'] = True
    else:
        if parser.options.start != None:
            updates['start'] = parser.options.start
        if parser.options.duration != None:
            updates['duration'] = parser.options.duration

    if parser.options.users != None:
        updates['users'] = None if parser.options.users == "*" else parser.options.users
    if parser.options.project != None:
        updates['project'] = parser.options.project
    if parser.options.cycle != None:
        updates['cycle'] = parser.options.cycle
    if not parser.no_args():
        updates['partitions'] = ":".join(parser.args)
    if parser.options.block_passthrough != None:
        updates['block_passthrough'] = parser.options.block_passthrough

    comp_args = ([{'name':parser.options.name}], updates, client_utils.getuid())
    client_utils.component_call(SCHMGR, False, 'set_reservations', comp_args)
    reservations = client_utils.component_call(SCHMGR, False, 'get_reservations', 
                                               ([{'name':parser.options.name, 'users':'*','start':'*', 'duration':'*', 'partitions':'*', 
                                                  'cycle': '*', 'res_id': '*', 'project':'*', 'block_passthrough':'*'}], ))
    client_utils.logger.info(reservations)
    client_utils.logger.info(client_utils.component_call(SCHMGR, False, 'check_reservations', ()))

def add_reservation(parser,spec,user):
    """
    add reservation 
    """
    validate_starttime(parser)

    spec['users']   = None if parser.options.users == '*' else parser.options.users
    spec['cycle']   = parser.options.cycle 
    spec['project'] = parser.options.project

    if parser.options.block_passthrough == None: spec['block_passthrough'] = False

    client_utils.logger.info(client_utils.component_call(SCHMGR, False, 'add_reservations', ([spec], user)))
    client_utils.logger.info(client_utils.component_call(SCHMGR, False, 'check_reservations', ()))

def main():
    """
    setres main
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    spec     = {} # map of destination option strings and parsed values
    opts     = {} # old map
    opt2spec = {}

    # list of callback with its arguments
    callbacks = [
        # <cb function>           <cb args>
        [ cb_time                , (False, True, True) ], # no delta time, Seconds, return int
        [ cb_date                , (True,) ], # Allow to set the date to 'now'
        [ cb_passthrough         , () ],
        [ cb_debug               , () ],
        [ cb_res_users           , () ]] 

    # Get the version information
    opt_def =  __doc__.replace('__revision__',__revision__)
    opt_def =  opt_def.replace('__version__',__version__)

    parser = ArgParse(opt_def,callbacks)

    user = client_utils.getuid()

    parser.parse_it() # parse the command line
    opt_count = client_utils.get_options(spec, opts, opt2spec, parser)

    # if continue to process options then
    if validate_args(parser,spec,opt_count):

        # modify an existing reservation
        if parser.options.modify_res != None:
            modify_reservation(parser)
        
        # add new reservation
        else:
            add_reservation(parser,spec,user)

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
